### ISSUE TITLE
Thread choice

### ISSUE THREAD

**Issue by glois** [chriskohlhoff/asio#68]
Greetings.

There seems to be a problem with how the thread implementation is chosen. In some cases, even though at least one type is possible, null_thread is selected. That will, of course, ruin everything.

Let's expose the important pieces. The first file is [thread.hpp](https://github.com/chriskohlhoff/asio/blob/master/asio/include/asio/detail/thread.hpp). The two big blocks have the same behavior : null_thread is always used (firstly included, secondly typedefed) when ASIO_HAS_THREADS is undefined. One could think that this is normal and that a lack of definition really means that there are no threads available. The rest of the blocks also give that impression because in case a definiton was found, it then proceeds to include the right files and typedef the right implementation. However, as it is now at least, ASIO_HAS_THREADS has little to do with the other implementations.

This is where the second file comes into play : [config.hpp](https://github.com/chriskohlhoff/asio/blob/asio-1-10-branch/asio/include/asio/detail/config.hpp#L841). This is where ASIO_HAS_THREADS is defined. Look at the conditions for yourself and compare them to those you can see in the blocks of the previous file.

In my case, the correct type would be std_thread (no boost, windows, posix/phtread for me). ASIO_HAS_STD_THREAD is correctly defined in config.hpp but it never has a chance to be used.

I believe the best fix would be to add this to the conditions in config.hpp : 

    #  elif defined(ASIO_HAS_STD_THREAD)
    #   define ASIO_HAS_THREADS 1

It fixes the problem for my particular case. I don't know if it is the only case or not and I haven't tested further than that. I hope this can be of help. Thanks for your hard work in any case.

**Comment by chriskohlhoff** [comment]
The definition of `ASIO_HAS_THREADS` is intended to mean something slightly different -- it's whether the compiler is building in multithreaded mode. E.g. for MSVC this means `_MT` is defined, for some unixes it means `_REENTRANT` etc. It's possible for a platform to provide the `<thread>` header but not be in multithreaded mode, which would result in an incorrect setting for `ASIO_HAS_THREADS`.

I need to clean up config.hpp's  definition of `ASIO_HAS_THREADS` to better reflect this. Looking at what boost does I should generally be testing for `_MT`, `__MT__`, `_REENTRANT` and so on.

Does your platform have the concept of a multithreaded mode, or is multithreading implicitly always on? If the former, does it use one of the commonly recognised macros. If the latter, what can I use to identify the platform? Thanks.

**Comment by glois** [comment]
Ah I guess there was a little misunderstanding then.
My platform is actually just Win7x64. I'm using mingw32 which doesn't provide an implementation for std::thread. For now, I decided to use [this little project](https://github.com/meganz/mingw-std-threads) to work around this issue. So __GNUC__ wouldn't really help but I haven't noticed any actual identifier for the project I linked, save for the guard definitions. It might be better to keep things as they are, then.

Just to be sure, when you say "building in multithreaded mode", that means "with a functional thread system", right? So basically with a working implementation behind the header? If so, the project I'm linking seems to respect the deal. I'm surprised that the opposite could happen on some platforms but I'll trust you on that. So having ASIO_HAS_STD_THREAD defined wouldn't be good enough. That said, I think that in a situation where <thread> is provided, I would personally have chosen to let ASIO use it and fail if it has to since it's the last step before using nullthread, which won't work anyway. It seems to me that it would be easier for someone in that case to understand his issue.

That basically mean that it's not really an issue for anyone but me which isn't a bad thing. I'll probably stick with my little fix for now when building with this setup.

Thanks for the clarification!
